src/details/ArborX_Ray.hpp

```diff
@@ -0,0 +1,182 @@
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+#ifndef ARBORX_RAY_HPP
+#define ARBORX_RAY_HPP
+
+#include <ArborX_Box.hpp>
+#include <ArborX_Point.hpp>
+
+#include <Kokkos_Macros.hpp>
+
+#include <cassert>
+#include <cmath>
+
+namespace ArborX
+{
+struct Ray
+{
+  using Vector = Point; // will regret this later
+
+  using Scalar = std::decay_t<decltype(std::declval<Vector>()[0])>;
+
+  class Normalized
+  {
+    Vector v_;
+
+  public:
+    KOKKOS_FUNCTION
+    explicit Normalized(Vector const &v)
+        : v_{v}
+    {
+    }
+
+    KOKKOS_FUNCTION
+    operator Vector &() { return v_; }
+
+    KOKKOS_FUNCTION
+    operator Vector const &() const { return v_; }
+  };
+
+  KOKKOS_DEFAULTED_FUNCTION
+  constexpr Ray() = default;
+
+  KOKKOS_FUNCTION
+  Ray(Point const &origin, Vector const &direction)
+      : _origin(origin)
+      , _direction(direction)
+  {
+    normalize(_direction);
+    for (int d = 0; d < 3; ++d)
+    {
+      // direction can be zero
+      _invdir[d] = 1.0 / (_direction[d]);
+    }
+  }
+
+  KOKKOS_FUNCTION
+  Ray(Point const &origin, Normalized const &direction)
+      : _origin(origin)
+      , _direction(direction)
+  {
+  }
+
+  KOKKOS_FUNCTION
+  static constexpr Scalar norm_sq(Vector const &v)
+  {
+    Scalar sq{};
+    for (int d = 0; d < 3; ++d)
+      sq += v[d] * v[d];
+    return sq;
+  }
+
+  KOKKOS_FUNCTION
+  static constexpr Scalar norm(Vector const &v)
+  {
+    return std::sqrt(norm_sq(v));
+  }
+
+  KOKKOS_FUNCTION static void normalize(Vector &v)
+  {
+    auto const n = norm(v);
+    assert(n > 0);
+    for (int d = 0; d < 3; ++d)
+      v[d] /= n;
+  }
+
+  KOKKOS_FUNCTION
+  constexpr Point &origin() { return _origin; }
+
+  KOKKOS_FUNCTION
+  constexpr Point const &origin() const { return _origin; }
+
+  KOKKOS_FUNCTION
+  constexpr Point &direction() { return _direction; }
+
+  KOKKOS_FUNCTION
+  constexpr Point const &direction() const { return _direction; }
+
+  KOKKOS_FUNCTION
+  constexpr Point &invdir() { return _invdir; }
+
+  KOKKOS_FUNCTION
+  constexpr Point const &invdir() const { return _invdir; }
+
+  Point _origin = {};
+  Point _direction = {{0.f, 0.f, 0.f}};
+  Point _invdir = {{0.f, 0.f, 0.f}};
+};
+
+namespace Details
+{
+namespace RayTracing
+{
+// ray-box intersection
+// The ray-box intersection algorithm is based on Majercik, A., et al. 2018.
+// Their 'efficient slag' algorithm checks the intersections both in front and
+// behind the ray. The function here checks the intersections in front of the
```

Comment:
This is not strictly correct. There algorithm does not work even in two directions, depending on how `min` and `max` functions are implemented.